kensu/pyspark/KensuDataFrameWriter.py

```diff
@@ -0,0 +1,229 @@
+        self._delayed_calls = []
+        self._options = {}
+        self._format = None
+
+    # FIXME: implement  __getattr__(self, item) for unknown field names to defer to self._df_writer
+
+    def _delay_fn_call(self, fn):
+        self._delayed_calls.append(fn)
+        return self
+
+    def _update_df(self, new_df):
+        self._df = new_df
+        self._df_writer = DataFrameWriter(new_df)
+
+    def _call_deferred_fns(self):
+        # always returns same self, so it's fine
+        for fn in self._delayed_calls:
+            fn()
+
+    @staticmethod
+    def _path_from_args(args, kwargs):
+        """
+        Allow to call methods:
+        - .parquet("abc.parquet") - args
+        - .parquet("abc.parquet", mode="overwrite") - args + kwargs
+        - .parquet(path="abc.parquet") - kwargs
+        - .parquet(mode="overwrite", path="abc.parquet") - kwargs in weird order - wouldn't work with `def parquet(path, *args, **kwrags)`
+
+        :return: path for first arg or from kwargs
+        """
+        return kwargs.get('path') or (args and args[0])
+
+    def mode(self, saveMode: Optional[str]) -> "KensuDataFrameWriter":
+        self._delay_fn_call(lambda: self._df_writer.mode(saveMode))
+        return self
+
+    def format(self, source: str) -> "KensuDataFrameWriter":
+        self._format = source
+        self._delay_fn_call(lambda: self._df_writer.format(source))
+        return self
+
+    def option(self, key: str, value: "OptionalPrimitiveType") -> "KensuDataFrameWriter":
+        self._options[key] = to_str(value)
+        self._delay_fn_call(lambda: self._df_writer.option(key, to_str(value)))
+        return self
+
+    def options(self, **options: "OptionalPrimitiveType") -> "KensuDataFrameWriter":
+        for k in options:
+            self._options[k] = to_str(options[k])
+        self._delay_fn_call(lambda: self._df_writer.options(**options))
+        return self
+
+    def _path_from_options(self):
+        return self._options.get('path')
+
+    @overload
+    def partitionBy(self, *cols: str) -> "KensuDataFrameWriter":
+        ...
+
+    @overload
+    def partitionBy(self, *cols: List[str]) -> "KensuDataFrameWriter":
+        ...
+
+    def partitionBy(self, *cols: Union[str, List[str]]) -> "KensuDataFrameWriter":
+        self._delay_fn_call(lambda: self._df_writer.partitionBy(*cols))
+        return self
+
+    @overload
+    def bucketBy(self, numBuckets: int, col: str, *cols: str) -> "KensuDataFrameWriter":
+        ...
+
+    @overload
+    def bucketBy(self, numBuckets: int, col: TupleOrListOfString) -> "KensuDataFrameWriter":
+        ...
+
+    def bucketBy(
+        self, numBuckets: int, col: Union[str, TupleOrListOfString], *cols: Optional[str]
+    ) -> "KensuDataFrameWriter":
+        self._delay_fn_call(lambda: self._df_writer.bucketBy(numBuckets, col, *cols))
+        return self
+
+    @overload
+    def sortBy(self, col: str, *cols: str) -> "KensuDataFrameWriter":
+        ...
```

Comment:
@vidma no need to implement those `...` ?

Reply:
it was like that in original pyspark code. `@overload` thingy.  this maybe might be python version dependent, and if needed might simplify it later